[dynamo][inductor] Provide public API to get compiler options/configs

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105026

issues resolved: https://github.com/pytorch/pytorch/issues/101832

**context**: get torch.compile config for further usage. E.g, the training platform wants to get if model is compiled with cudagraph enabled and trigger further action

**how it is implemented**
   * the core logic is backend.get_compiler_config() in torch/_dynamo/eval_frame.py
   * for backend='inductor' / _TorchCompileInductorWrapper, we have inductor-specific implementation in get_compiler_config in torch/_inductor/compile_fx.py and torch/__init__.py

**how to use it**: Below is an example.

```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": True}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

weifengpy committed:
[dynamo][inductor] Provide public API to get compiler options/configs

issues resolved: https://github.com/pytorch/pytorch/issues/101832

context: get torch.compile config for further usage. E.g,
the training platform wants to get if model is compiled
with cudagraph enabled and trigger further action

this PR implemented OptimizedModule.get_compiler_config(). Example usage
```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": not default_cudagraphs}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo][inductor] Provide public API to get compiler options/configs"


issues resolved: https://github.com/pytorch/pytorch/issues/101832

**context**: get torch.compile config for further usage. E.g, the training platform wants to get if model is compiled with cudagraph enabled and trigger further action

**How to use it**: this PR implemented OptimizedModule.get_compiler_config(). Below is an example.

**limitation**: get_compiler_config only works for torch.nn.Module instead of function. For function, torch.compile returns a function, it requires more disruptive code change

```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": True}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo][inductor] Provide public API to get compiler options/configs"


issues resolved: https://github.com/pytorch/pytorch/issues/101832

**context**: get torch.compile config for further usage. E.g, the training platform wants to get if model is compiled with cudagraph enabled and trigger further action

**How to use it**: this PR implemented OptimizedModule.get_compiler_config(). Below is an example.

**limitation**: get_compiler_config only works for torch.nn.Module instead of function. For function, torch.compile returns a function, it requires more disruptive code change

```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": True}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

[ghstack-poisoned]

yanboliang reviewed (approved):
Looks good to me! cc @jansel for final review.

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -148,6 +149,13 @@ def __dir__(self):
             attr for attr in super().__dir__() if attr not in orig_mod_attrs
         ]
 
+    def get_compiler_config(self):
```
Inductor-specific code shouldn't go in `/_dynamo/`, we should try to keep things backend agnostic.

willfengg replied:
Got you. I updated the diff with ```backend.get_compiler_config()``` so it becomes more backend agnostic

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -148,6 +149,13 @@ def __dir__(self):
+        from torch._inductor import config as inductor_config
+
+        with inductor_config.patch(self.dynamo_ctx.config_patches):
+            return copy.deepcopy(inductor_config._config)
```
Add a helper method to config_utils.py rather than reading private attribute.

willfengg replied:
added helper method get_config_copy in torch/_dynamo/config_utils.py for reading private attribute

weifengpy committed:
Update on "[dynamo][inductor] Provide public API to get compiler options/configs"


issues resolved: https://github.com/pytorch/pytorch/issues/101832

**context**: get torch.compile config for further usage. E.g, the training platform wants to get if model is compiled with cudagraph enabled and trigger further action

**How to use it**: this PR implemented OptimizedModule.get_compiler_config(). Below is an example.

**limitation**: get_compiler_config only works for torch.nn.Module instead of function. For function, torch.compile returns a function, it requires more disruptive code change for provide get_compiler_config() api

```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": True}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo][inductor] Provide public API to get compiler options/configs"


issues resolved: https://github.com/pytorch/pytorch/issues/101832

**context**: get torch.compile config for further usage. E.g, the training platform wants to get if model is compiled with cudagraph enabled and trigger further action

**How to use it**: this PR implemented OptimizedModule.get_compiler_config(). Below is an example.

**limitation**: get_compiler_config only works for torch.nn.Module instead of function. For function, torch.compile returns a function, it requires more disruptive code change for provide get_compiler_config() api

```
model = DummyModule()
optimized_module = torch.compile(
    model, options={"triton.cudagraphs": True}
)
compiler_config = optimized_module.get_compiler_config()

if compiler_config["triton.cudagraphs"]:
   pass
```

[ghstack-poisoned]

willfengg commented:
@pytorchbot merge
